ReactAndroid/src/main/java/com/facebook/react/devsupport/DevServerHelper.java

```diff
@@ -316,45 +325,109 @@ public void onFailure(Call call, IOException e) {
       }
 
       @Override
-      public void onResponse(Call call, Response response) throws IOException {
+      public void onResponse(Call call, final Response response) throws IOException {
         // ignore callback if call was cancelled
         if (mDownloadBundleFromURLCall == null || mDownloadBundleFromURLCall.isCanceled()) {
           mDownloadBundleFromURLCall = null;
           return;
         }
         mDownloadBundleFromURLCall = null;
 
-        // Check for server errors. If the server error has the expected form, fail with more info.
-        if (!response.isSuccessful()) {
-          String body = response.body().string();
-          DebugServerException debugServerException = DebugServerException.parse(body);
-          if (debugServerException != null) {
-            callback.onFailure(debugServerException);
-          } else {
-            StringBuilder sb = new StringBuilder();
-            sb.append("The development server returned response error code: ").append(response.code()).append("\n\n")
-              .append("URL: ").append(call.request().url().toString()).append("\n\n")
-              .append("Body:\n")
-              .append(body);
-            callback.onFailure(new DebugServerException(sb.toString()));
-          }
-          return;
-        }
+        final String url = response.request().url().toString();
 
-        Sink output = null;
-        try {
-          output = Okio.sink(outputFile);
-          Okio.buffer(response.body().source()).readAll(output);
-          callback.onSuccess();
-        } finally {
-          if (output != null) {
-            output.close();
+        // Make sure the result is a multipart response and parse the boundary.
+        String contentType = response.header("content-type");
+        Pattern regex = Pattern.compile("multipart/mixed;.*boundary=\"([^\"]+)\"");
+        Matcher match = regex.matcher(contentType);
+        if (match.find()) {
+          String boundary = match.group(1);
+          MultipartStreamReader bodyReader = new MultipartStreamReader(response.body().source(), boundary);
+          boolean completed = bodyReader.readAllParts(new MultipartStreamReader.ChunkCallback() {
+            @Override
+            public void execute(Map<String, String> headers, Buffer body, boolean finished) throws IOException {
+              // This will get executed for every chunk of the multipart response. The last chunk
+              // (finished = true) will be the JS bundle, the other ones will be progress events
+              // encoded as JSON.
+              if (finished) {
+                // The http status code for each separate chunk is in the X-Http-Status header.
+                int status = response.code();
```

Comment:
Just in case the chunk doesn't specify a status we use the one from the request.